### PR TITLE
Display location of comments in activity stream.

### PR DESCRIPTION
**Pull request by shurcooL**
Previously, comment events would only include the repository. Now, it also displays the issue, PR, or commit that the comment is located in.

This provides more context to the comment and makes things more readable. It also provides a link to the real comment for full details and context.

Fetch additional commits from `github.CommitCommentEvent` events.

Factor out details aspect from various components to higher level. This way, they can be used for details and for inline spans.

### Before

image

### After

image

### CODE REVIEW

File: index.go [hunk 1]
```diff
@@ -633,61 +725,78 @@ border-radius: 3px;`
...
 func shortSHA(sha string) string {
 	return sha[:8]
 }
 
+func shortCommit(s string) string {
+	if len(s) <= 24 {
+		return s
+	}
+	return s[:23] + "…"
+}
```
Comment (bradleyfalzon) [inline index.go]: Perhaps `func ellipse(s string, len int) string` to replace `shortCommit` and `shortTitle`?

File: index.go [hunk 2]
```diff
@@ -471,6 +491,78 @@ func (a activity) Render() []*html.Node {
...
+		URL:     *p.Comment.HTMLURL,
+		Black:   true,
+	}
+	switch *p.Issue.State {
+	case "open":
+		n.Icon = octiconssvg.GitPullRequest
+		n.Color = RGB{R: 0x6c, G: 0xc6, B: 0x44} // Green.
+	case "closed":
+		n.Icon = octiconssvg.GitPullRequest
+		n.Color = RGB{R: 0xbd, G: 0x2c, B: 0x00} // Red.
```
Comment (shurcooL) [inline index.go]: Thanks for the suggestion. I have considered it, and I know with high certainty that this is a refactor that's not worth doing at this time. The benefits are very little.

The reason is that I have these red/green/purple colors indicating merge states (that map to GitHub colors) in quite a few places in my GOPATH. However, some of them use [`issues.RGB`](https://godoc.org/github.com/shurcooL/issues#RGB) type, some use [`notifications.RGB`](https://godoc.org/github.com/shurcooL/notifications#RGB). I have the `RGB` color type defined locally in a few places, including in this repo.

```
Searching 1427 files for "RGB{R: 0xbd, G: 0x2c, B: 0x00}"

/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/shurcooL/home/index.go:
  278  			case "closed":
  279  				details.Icon = octiconssvg.IssueClosed
  280: 				details.Color = RGB{R: 0xbd, G: 0x2c, B: 0x00} // Red.
  281  			case "reopened":
  282  				details.Icon = octiconssvg.IssueReopened
  ...
  306  				e.Action = component.Text("closed a pull request in")
  307  				details.Icon = octiconssvg.GitPullRequest
  308: 				details.Color = RGB{R: 0xbd, G: 0x2c, B: 0x00} // Red.
  309  			case *p.PullRequest.Merged:
  310  				e.Action = component.Text("merged a pull request in")
  ...
  505  	case "closed":
  506  		n.Icon = octiconssvg.IssueClosed
  507: 		n.Color = RGB{R: 0xbd, G: 0x2c, B: 0x00} // Red.
  508  	default:
  509  		log.Println("activity.Render: unsupported *github.IssueCommentEvent Issue.State:", *p.Issue.State)
  ...
  525  	case "closed":
  526  		n.Icon = octiconssvg.GitPullRequest
  527: 		n.Color = RGB{R: 0xbd, G: 0x2c, B: 0x00} // Red.
  528  	// TODO: Detect merged somehow? It's likely going to require making an API call.
  529  	default:
  ...
  546  	case p.PullRequest.MergedAt == nil && *p.PullRequest.State == "closed":
  547  		n.Icon = octiconssvg.GitPullRequest
  548: 		n.Color = RGB{R: 0xbd, G: 0x2c, B: 0x00} // Red.
  549  	case p.PullRequest.MergedAt != nil:
  550  		n.Icon = octiconssvg.GitMerge

/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/shurcooL/issues/fs/notifications.go:
  120  		return notifications.RGB{R: 0x6c, G: 0xc6, B: 0x44}
  121  	case issues.ClosedState: // Closed.
  122: 		return notifications.RGB{R: 0xbd, G: 0x2c, B: 0x00}
  123  	default:
  124  		return notifications.RGB{}

/Users/Dmitri/Dropbox/Work/2013/GoLand/src/github.com/shurcooL/notifications/githubapi/githubapi.go:
   83  			case err == nil && state == "closed":
   84  				notification.Icon = "issue-closed"
   85: 				notification.Color = notifications.RGB{R: 0xbd, G: 0x2c, B: 0x00} // Red.
   86  			default:
   87  				notification.Icon = "issue-opened"
   ..
  102  				notification.Color = notifications.RGB{R: 0x6c, G: 0xc6, B: 0x44} // Green.
  103  			case err == nil && state == "closed":
  104: 				notification.Color = notifications.RGB{R: 0xbd, G: 0x2c, B: 0x00} // Red.
  105  			case err == nil && state == "merged":
  106  				notification.Color = notifications.RGB{R: 0x6e, G: 0x54, B: 0x94} // Purple.

8 matches across 3 files
```

It's very easy to find all instances of these colors and keep them in sync. They change very infrequently (almost never).

Until I find a good place for that `RGB` type, I couldn't have a globally usable set of colors, I'd need to cast things anyway... And coming up with a new repo for `RGB` is a lot of work, etc.

So in the end, this is something I've thought about and will deal with later, it's not causing any downsides to leave it as is for now. Thanks for the idea though!

File: index.go (hunk 1 above)
Comment (shurcooL) [inline index.go]: That would make it harder to adjust the 4 types of lengths I use. It's currently 200 for bodies, 36 for title, 24 for commit message, and 8 for SHAs. That corresponds to the 4 funcs.

If I used `ellipsis(s string, l int)` everywhere, it'd be harder to update those 4 categories.

I _could_ rewrite `shortCommit`, `shortBody` and `shortTitle` to use `ellipsis(s string, l int)`, but it doesn't seem to simplify the code that much:

```diff
diff --git a/index.go b/index.go
index 1ce36b1..5b7cb13 100644
--- a/index.go
+++ b/index.go
@@ -694,19 +694,9 @@ func (d imageText) Render() []*html.Node {
 	return []*html.Node{div}
 }
 
-func shortBody(s string) string {
-	if len(s) <= 200 {
-		return s
-	}
-	return s[:199] + "…"
-}
+func shortBody(s string) string { return ellipsis(s, 200) }
 
-func shortTitle(s string) string {
-	if len(s) <= 36 {
-		return s
-	}
-	return s[:35] + "…"
-}
+func shortTitle(s string) string { return ellipsis(s, 36) }
 
 type code struct {
 	Text          string
@@ -790,12 +780,7 @@ func shortSHA(sha string) string {
 	return sha[:8]
 }
 
-func shortCommit(s string) string {
-	if len(s) <= 24 {
-		return s
-	}
-	return s[:23] + "…"
-}
+func shortCommit(s string) string { return ellipsis(s, 24) }
 
 // firstParagraph returns the first paragraph of text s.
 func firstParagraph(s string) string {
@@ -854,3 +839,10 @@ func (d pages) Render() []*html.Node {
 
 	return nodes
 }
+
+func ellipsis(s string, l int) string {
+	if len(s) <= l {
+		return s
+	}
+	return s[:l-1] + "…"
+}
```

There's only 3 funcs, they're very short and easy to read, modify. And I expect they may change or go away soon. That's why I don't think trying to generalize them right now would buy a lot.

Thanks for the suggestion!

**Comment by shurcooL** [comment]
> review of review appreciated

@bradleyfalzon Thanks for looking, appreciated! The fact that you haven't found much more serious issues is a great thing.

I will take comments into consideration. I replied to give you feedback on them.

**Pull request merged**